Add MSYS2 windows build workflow

shmup opened the pull request:
This adds a windows build to both release and nightly. You can check it out on my fork:

https://github.com/shmup/crankboy-app/releases/tag/untagged-0b7bb87e1fe9bfb75d85

Though I'll admit I don't have Windows and haven't ran the artifact!

fixes #43

shmup committed:
Add MSYS2 windows build workflow

stonerl commented:
Wow, thanks allot. 🙏 I will test this later. I would add the Windows SDK to our toolchain repo later. Would you kindly adapt the script to make use of the submodule after I've done so?

stonerl commented:
@shmup I added the SDK to the toolchain repo. Would you kindly try with that?

shmup commented:
Not a problem @stonerl , I've rebased and triggered another release to build: https://github.com/shmup/crankboy-app/releases/tag/v1.0.5

stonerl commented:
Sorry for getting back so late. I downloaded the artifact the workflow is creating, and it is an `.exe` file, but it should be a `.pdx`. Or at least a `.dll` if you only building the lib. Not sure what is going on there; would you kindly check?

You can take a look at the sim builds for Mac or Linux to see what the final build should look like?. And again thank you for trieng to get this work.

maxoakland commented:
> This adds a windows build to both release and nightly. You can check it out on my fork:
> 
> https://github.com/shmup/crankboy-app/releases/tag/untagged-0b7bb87e1fe9bfb75d85
> 
> Though I'll admit I don't have Windows and haven't ran the artifact!
> 
> fixes #43

That link is broken

shmup commented:
yeah @maxoakland , that is for a draft release, not realizing those aren't viewable. I had to fix it anyway, here ya go @stonerl https://github.com/shmup/crankboy-app/releases/tag/v1.0.14

let me know if anything else should be addressed, cheers

nstbayless committed:
WIP CGB support

# Conflicts:
#	libs/peanut_gb.h

# Conflicts:
#	libs/peanut_gb.h
#	libs/pgb/pgb_v2.h

nstbayless committed:
added 'experimental CGB mode' to options

nstbayless committed:
CGB mode option

nstbayless committed:
fixed settings crash

nstbayless committed:
multiple itcm cores

nstbayless committed:
first differences between cgb and dmg core

nstbayless committed:
add draw_line for cgb (bg and window)

nstbayless committed:
swap itcm/dtcm order

nstbayless committed:
removed prefetch for tile data

nstbayless committed:
serial commands

# Conflicts:
#	src/app.c

nstbayless committed:
fixed itcm

nstbayless committed:
cgb sprite banks

nstbayless committed:
fixed fatal error in cgb initialization after prompt

stonerl commented:
@shmup Awesome. May I ask you for one additional favor? Currently the simulator build only contains the `.dll`. The idea of the simulator build is also that users can transfer the app easily to their device. For this to work, we also need the `pdx.bin`, the executable for the Playdate. Can you add this as well? This would be very nice of you.

I already added the ARM toolchain to the submodule.

nstbayless committed:
added audio to pgb version headers; fixed save state loading problem

shmup commented:
@stonerl can do! I'll hack on it this week

nstbayless committed:
experimental: lock button override

nstbayless committed:
/Shared/Emulation/gb

nstbayless committed:
.fosl install

nstbayless committed:
update lock button warning

stonerl committed:
refactor(ui): Adopt new enum for lock button warning modal

Updates the warning modal for the lock button override feature to use the new `CB_ModalWarningPosition` enum.

This commit replaces the previous boolean assignment (`modal->warning = true;`) with the more explicit and descriptive `modal->warning = CB_MODAL_WARNING_TOP;`.

Signed-off-by: Toni Förster <toni.foerster@icloud.com>

nstbayless committed:
fos-package

nstbayless committed:
utility update

nstbayless committed:
update fos-package

nstbayless committed:
added fos asset to assets

stonerl commented:
I resolved the merge conflicts. I am not sure why GH decided to add a diff for `settings.c` but I will take care of this later.

stonerl commented:
Okay, I completely rebased and applied your patches again.

nstbayless committed:
check rom hash when downloading patch

nstbayless committed:
start checking for update earlier, on boot

nstbayless committed:
removed incorrect 'cgb' field from pgb_v1.h

nstbayless committed:
fixed performance.

nstbayless committed:
renamed uncapped fps -> turbo speed

nstbayless committed:
first-time launch intro

stonerl committed:
feat(script): Add battery removal and refactor patching logic

This commit introduces a new feature to the ROM patcher script, allowing users to remove the battery-backed save (SRAM) capability from a Game Boy ROM. It also refactors the existing code for better modularity and extensibility.

The new 'battery' patch works by modifying the cartridge type byte at address 0x0147 to its non-battery equivalent, based on the Pan Docs specification. This is a common requirement for users of certain flash carts or for specific ROM hacking projects.

Key changes include:

- New 'battery' patch option: A user can now specify `battery` as a command-line argument to patch the cartridge type.
- Cartridge Type Mapping: A dictionary `BATTERY_CARTRIDGE_TYPES` has been added to map battery-inclusive hardware types to their non-battery counterparts.
- Command-Line Argument Parsing: The script now accepts a second argument to determine which patch to apply (`cgb` or `battery`).
- Code Refactoring:
    - The CGB flag patching, battery removal, and checksum recalculation logic have been moved into their own dedicated functions (`patch_cgb_flag`, `remove_battery`, `update_checksums`).
    - This improves readability and makes the script easier to maintain and extend with new patch types in the future.
- Improved User Feedback: The script now provides more specific output based on the patch being applied.

Signed-off-by: Toni Förster <toni.foerster@icloud.com>

nstbayless committed:
make HTTP api safe, guarantees cb invocation

shmup commented:
apologies for the delay ya'll. any day now.. (just started a new gig)

nstbayless committed:
WIP homebrew hub support

nstbayless committed:
fixed cycle count error in opcode 0xf8

nstbayless committed:
homebrew hub updates

nstbayless committed:
homebrew downloading implemented

nstbayless committed:
update version to 1.2

nstbayless committed:
'Nightly' in credits version

nstbayless committed:
skip http-enable if already enabled

nstbayless committed:
improvements to homebrew downloading

nstbayless committed:
use strrchr for get_extension instead of strchr

nstbayless committed:
homebrew hub scene: prompt to quit if system privileges

nstbayless committed:
don't save homebrew.pdi if no image to be drawn

nstbayless committed:
hbh: wait prompt

nstbayless committed:
use playdate->system->restart

nstbayless committed:
script: metroid 2 (wip)

nstbayless committed:
script: metroid 2 (wip/2)

nstbayless committed:
metroid 2 HUD

nstbayless committed:
save map exploration

nstbayless committed:
script: metroid 2, HUD polish

nstbayless committed:
script ready: metroid 2

nstbayless committed:
script: metroid: beam rooms

nstbayless committed:
script: metroid 2: hex map

nstbayless committed:
script: metroid 2: fix audio bug

nstbayless committed:
script: metroid 2: polish

nstbayless committed:
replace CGB prompt

nstbayless committed:
library setting to prompt for cgb-optional ROMs

nstbayless committed:
changed description for CGB prompt option

nstbayless committed:
update README

nstbayless committed:
different save bin for ROM hacks

nstbayless committed:
scripts credits update

nstbayless committed:
metroid 2: crank weapon select fix

nstbayless committed:
save data slot

nstbayless committed:
disable auto-lock during patch downloading scene and homebrew hub scene

nstbayless committed:
update script option description

nstbayless committed:
removed unused scripts

nstbayless committed:
warn if save data has different patch set

nstbayless committed:
adjust settings description for save slot

nstbayless committed:
fix mac build

nstbayless committed:
fixed http safe crash and patch set hash reading

nstbayless committed:
update info for met2 script

nstbayless committed:
fixed crash in metroid 2 map in certain areas

nstbayless committed:
fix apple build

nstbayless committed:
'library' options and 'ui sounds' setting always appears in library view settings

nstbayless committed:
renamed PREFBITS_LIBRARY_ONLY to PREFBITS_ALWAYS_GLOBAL

nstbayless committed:
reset save slot to 0 when exiting game scene

nstbayless committed:
fixed app duplicate field

nstbayless committed:
script: metroid 2: update map

nstbayless committed:
metroid 2: force crank dock/undock off

nstbayless committed:
fixed preference saving

nstbayless committed:
parental lock

nstbayless committed:
fix met2 hp

Pomdap commented:
I tested the build for the simulator in a Windows environment.
The only issue was a missing dependency on the library libwinpthread-1.dll.
I resolved the problem by adding the -static argument to the gcc command, it might be worth adding this to the scripts.

shmup committed:
Exclude sys/wait in windows build

shmup commented:
Okaaay, I did some things!

Here's the latest release from my development branch:

https://github.com/shmup/crankboy-app/releases/tag/v1.0.16

@stonerl let me know how this looks now

@Pomdap cool good looking out, I've went ahead and added that to this PR for release/nightly workflows. Mind trying again?

I also added a new commit that resolves something introduced recently that isn't available in windows, see: https://github.com/shmup/crankboy-app/commit/8a7ea42dbe182af97b013d2653b5bc6ea2854c0a

Pomdap commented:
I had to clean the simulator directory to fix a crash during loading, but it works now I can launch the app in the simulator.

I also tried sideloading the app through the website, but I get the following error: "_Could not find a top-level .pdx file_".
It looks like the CrankBoy-Windows folder needs to be removed during the package creation, maybe by doing something like:

```
- name: Package Windows Simulator Build
  run: |
      zip -r CrankBoy-Simulator-Windows.zip CrankBoy.pdx
```


I also tried uploading it through the simulator. Installation works, but launching doesn’t (same issue if you sideload through the web with a corrected zip).
The following error appears on the Playdate screen: "_no such function 'update'_"

Thank you for your work 👍, I hope this information is helpful.

shmup commented:
Thanks again @Pomdap , I've attempted resolving both outcomes. I'd like to test Windows myself here soon so I don't need this feedback loop :P 

If you get around to testing before me, I made the changes under the .16 release

Pomdap commented:
I was able to run a game with the Windows package on the simulator, and the playdate by loading it directly from the simulator and by using the sideload function on the website, nice job. 👍🏻

shmup commented:
Whenever this PR is ready, I'd like to rebase and reduce its commits to some extent, just FYI @stonerl

stonerl commented:
@shmup no worries. Just squash your commits when this is ready to being merged and I'll be rebasing it on dev HEAD.

shmup commented:
Good to go @stonerl

stonerl committed:
Merge branch 'development' into development

stonerl commented:
@shmup can I add you by your full name to our credits list?

shmup commented:
Perfectly fine with me, @stonerl ! Cheers
